### PR TITLE
Feature/nsa 8412 manage console myesf grant types

### PR DESCRIPTION
**Pull request by AMcCartneyMTH**
Bug fix: Changes have been made to address an issue where the grant types were not being set correctly for the hybrid auth flow. As part of this change, the test pack has been refactored to better cover the grant type logic and become more readable.

Related ticket: [NSA-8412](https://dfe-secureaccess.atlassian.net/browse/NSA-8412)

### CODE REVIEW

File: src/app/services/serviceConfig.js [hunk 1]
```diff
@@ -3,43 +3,22 @@ const he = require("he");
...
   const { CLIENT_SECRET_POST, CLIENT_SECRET_BASIC } = TOKEN_ENDPOINT_AUTH_METHOD;
   const sessionValue = sessionService?.tokenEndpointAuthMethod?.newValue;
   const fallbackValue = service.relyingParty.token_endpoint_auth_method === CLIENT_SECRET_POST ? CLIENT_SECRET_POST : CLIENT_SECRET_BASIC;
 
-  const responseTypes = (sessionService?.responseTypes?.newValue || service.relyingParty.response_types) || [];
+  const responseTypes = sessionService?.responseTypes?.newValue || service.relyingParty.response_types || [];
   const authFlowType = determineAuthFlowByRespType(responseTypes);
 
-  tokenEndpointAuthMethod = (sessionValue && (authFlowType === AUTHENTICATION_FLOWS.HYBRID_FLOW || authFlowType === AUTHENTICATION_FLOWS.AUTHORISATION_CODE_FLOW)) ? sessionValue : fallbackValue;
+  tokenEndpointAuthMethod = sessionValue && (authFlowType === AUTHENTICATION_FLOWS.HYBRID_FLOW || authFlowType === AUTHENTICATION_FLOWS.AUTHORISATION_CODE_FLOW) ? sessionValue : fallbackValue;
```
Comment (insacuri) [inline src/app/services/serviceConfig.js:21]: Small suggestion (so small that you can opt into it if you want!), but maybe you could make a `isAuthFlowHybridOrAuthorisationCode` (or something like that) function, so then this line turns into 

```
const isAuthFlowHybridOrAuthorisationCode = isAuthFlowHybridOrAuthorisationCode(authFlow);
tokenEndpointAuthMethod = sessionValue && isAuthFlowHybridOrAuthorisationCode ? sessionValue : fallbackValue;
```

I always try to make ternary operators as easy to read as possible and I think that helps?  But maybe it's too wordy?  Your call ^^

Comment (insacuri) [reply src/app/services/serviceConfig.js:21]: Or maybe reuse the
```
const isAuthorisationCodeFlow = authFlowType === AUTHENTICATION_FLOWS.AUTHORISATION_CODE_FLOW;`
``` 
code from further down in this file.  Something like
```
tokenEndpointAuthMethod = sessionValue && (isAuthorisationCodeFlow || isHybridFlow) ? sessionValue : fallbackValue;
```
is easier to read than the original.

File: src/app/services/serviceConfig.js [hunk 2]
```diff
@@ -391,9 +378,13 @@ const postServiceConfig = async (req, res) => {
   try {
     const serviceModels = await buildCurrentServiceModel(req);
 
+    // const currentService = serviceModels.currentServiceModel;
```
Comment (insacuri) [inline src/app/services/serviceConfig.js:381]: Remove commented out code if it's not going to be used again.

File: test/app/services/postServiceConfig.test.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-const { urlValidator } = require("login.dfe.validation/src/urlValidator");
+//const { urlValidator } = require("login.dfe.validation/src/urlValidator");
```
Comment (insacuri) [inline test/app/services/postServiceConfig.test.js:1]: Same comment again about commented out code

File: test/app/services/postServiceConfig.test.js [hunk 4]
```diff
@@ -115,69 +203,31 @@ describe("when editing the service configuration", () => {
...
-
+  // COMMON Tests
   it("then it should render view without validation if service home url is an empty string", async () => {
+    // ARRANGE
     req.body.serviceHome = "";
 
+    // ACT
     await postServiceConfig(req, res);
+
+    // ACT
```
Comment (insacuri) [inline test/app/services/postServiceConfig.test.js:218]: Should this be ASSERT?

File: test/app/services/serviceConfig.test.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+describe('when determining the grant types to set', () => {
+    it('should return authorization code grant type with refresh token for hybrid or authorization code flow', () => {
+        const result = getGrantTypes({
+            isHybridFlow: true,
+            isAuthorisationCodeFlow: false,
+            isImplicitFlow: false,
+            refreshToken: 'refresh_token',
+            oldService: null,
+        });
+        // expect(result).toEqual(['authorization_code', 'refresh_token']);
```
Comment (insacuri) [inline test/app/services/serviceConfig.test.js:16]: Can this be removed if not needed?

**Review by insacuri — CHANGES_REQUESTED** [review]
Changes look generally good.  Just a few nitpicky bits and it should be good to go!

File: test/app/services/postServiceConfig.test.js (hunk 3 above)
Comment (AMcCartneyMTH) [reply test/app/services/postServiceConfig.test.js:1]: Good spot. Thanks. Had commented this out to make sure it wasn't being required, but didn't delete it.

File: test/app/services/serviceConfig.test.js (hunk 5 above)
Comment (AMcCartneyMTH) [reply test/app/services/serviceConfig.test.js:16]: Certainly.

File: test/app/services/postServiceConfig.test.js (hunk 4 above)
Comment (AMcCartneyMTH) [reply test/app/services/postServiceConfig.test.js:218]: It should - I got carried away typing // ACT by this point - thank you :)

File: src/app/services/serviceConfig.js (hunk 2 above)
Comment (AMcCartneyMTH) [reply src/app/services/serviceConfig.js:381]: Removed.

**Review by insacuri — DISMISSED** [review]

File: src/app/services/serviceConfig.js (hunk 1 above)
Comment (AMcCartneyMTH) [reply src/app/services/serviceConfig.js]: I think this is a good point and certainly helps with readability. I've made this change.

**Review by insacuri — APPROVED** [review]

**Review by trickyward — APPROVED** [review]
Looking over the code, its looks good to me.
